### DISCUSSION TITLE
Unable to run bats tests, commands not found

### DISCUSSION THREAD

**Discussion by davehouser1** [bats-core/bats-core#669] [Q&A]
Hello, 

I was notified that our bats tests were not able to start recently. We get these errors:

```
 sudo bats ./ourtests.bats
/usr/local/libexec/bats-core/bats: line 98: bats_readlinkf: command not found
/usr/local/libexec/bats-core/bats: line 464: exec: bats-exec-suite: not found
/usr/local/libexec/bats-core/bats: line 466: bats-format-pretty: command not found
```

Not sure what is causing this. I was told that bats-core was upgraded to 1.8 on the host that is running the tests recently. It was last working on 1.5, would that cause these issues? Anyway to fix this issue? would you recommend we downgrade to 1.5 again? Is there a downgrade process to follow, or just download / clone the branch for 1.5, and run `install.sh` again?

**Comment by martin-schulze-vireso** [comment]
This reads like BATS_ROOT is not set correctly, which is set by bin/bats. Did you accidentally add libexec/bats-core/bats to your PATH so `sudo bats` runs the wrong executable?

**Reply by davehouser1** [reply]
I just cracked open /bin/bats, here is what is at the end:

```
export BATS_ROOT
bats_resolve_absolute_root_dir "$0" 'BATS_ROOT'
exec "$BATS_ROOT/libexec/bats-core/bats" "$@"
```

Running `echo $BATS_ROOT` shows nothing. Should it be set to something?

**Reply by martin-schulze-vireso** [reply]
Do I read that correctly: you have /bin/bats and /usr/local/libexec/bats-core/bats?

**Reply by davehouser1** [reply]
Yes, I appear to have both.

**Reply by martin-schulze-vireso** [reply]
You should have two bats executables, but they should have the same prefix, whereas in your case one hast prefix / and the other /usr/libexec.

This leads to problems, as bin/bats uses its prefix to infer the location of libexec/bats-core/bats.

However, there seems to be more going on, since the error messages don't fit exactly in this pattern. What is you PATH?

**Reply by davehouser1** [reply]
Here is the user's path running with and without sudo
```
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
Here is root's path
```
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
```

**Reply by martin-schulze-vireso** [reply]
> You should have two bats executables, but they should have the same prefix, whereas in your case one hast prefix / and the other /usr/libexec

I meant /usr (for libexec/bats-core/bats).

**Reply by martin-schulze-vireso** [reply]
> > You should have two bats executables, but they should have the same prefix, whereas in your case one hast prefix / and the other /usr/libexec
> 
> I meant /usr (for libexec/bats-core/bats). 

Still wrong its /usr/local

Can you do a search of all files named bats? Maybe you have two installations mixing...

**Reply by davehouser1** [reply]
Here is the output:

```
# find / -name "bats" -type f 2>/dev/null
/usr/local/bin/bats-core/bats
/usr/local/bin/bats
/usr/local/libexec/bats-core/bats
/usr/libexec/bats-core/bats
```

Huh, when running the following it seems to work:
```
/usr/local/bin/bats --version
Bats 1.8.0
```

ok seems like we need to clean up other versions of bats on here. Will look into this. Thanks for the path help @martin-schulze-vireso

**Reply by martin-schulze-vireso** [reply]
> ```
> # find / -name "bats" -type f 2>/dev/null
> /usr/local/bin/bats-core/bats
> /usr/local/bin/bats
> /usr/local/libexec/bats-core/bats
> /usr/libexec/bats-core/bats
> ``` 

That /usr/local/bin/bats-core/bats looks very wrong and /usr/libexec/bats-core/bats should imply /usr/bin/bats as well.
